### PR TITLE
[react-router] Improve withRouter definition with conditional types

### LINKED ISSUE #24070: react-router: `withRouter` narrows components with `any` prop type to `{}`
Given a component that has a props type of `any`, when wrapping the component in the `withRouter` HOC, the returned component has a prop type of `{}`. I would expect the return component to have a prop type of `any`.

`any` should not be narrowed to `{}`.

Components with `any` type are common when gradually migrating a codebase to TypeScript, where many components may have their prop types undefined whilst the codebase is gradually migrated.

Example:

``` tsx
import React, { SFC } from 'react';
import { withRouter } from 'react-router-dom';

const MyComponent: SFC<any> = () => null;

const MyComponentWithRouter = withRouter(MyComponent);

<MyComponentWithRouter
  // Unexpected error: Property 'foo' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Pick<any, never>, ComponentState>> & Rea...'.
  foo={1}
/>
```

Related https://github.com/gcanti/typelevel-ts/issues/19

This issue was introduced by the change in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21329 /cc @Grmiade (PR author)

---

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/react-router` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [ ] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @sergey-buturlakin @mrk21 @vasek17 @ngbrown @awendland @KostyaEsmukov @johnnyreilly @LKay @DovydasNavickas @tkrotoff @huy-nguyen @grmiade @DaIgeb @egorshulga @neuoy 

If you do not mention the authors the issue will be ignored.

### PR DESCRIPTION
**Pull request by Grmiade**
⚠️ We have to wait for a compatibility with the new 2.8 TypeScript version

Thanks to conditional types, withRouter definition excludes injected props only when props is assignable to RouteComponentProps.

Fix #24070

### CODE REVIEW

**Comment by Grmiade** [comment]
@johnnyreilly I have this error `react-router depends on react but has a lower required TypeScript version.`
But if I change the TypeScript version, I will have to change the version in all packages which depends on `react` 😱 What should I do?

**Comment by johnnyreilly** [comment]
I think that may be the only course of action I'm afraid.  It's not great..

**Comment by Grmiade** [comment]
screen shot 2018-03-28 at 15 06 52
😓

**Comment by johnnyreilly** [comment]
Ouch

**Comment by Grmiade** [comment]
`Error: Could not parse version: line is '// TypeScript Version: 2.8'`
How resolve this error? We should only update the `definitelytyped-header-parser` package ?

**Comment by DanielRosenwasser** [comment]
Because of the fact that this would break literally everyone who hasn't upgraded to 2.8, DefinitelyTyped is only compatible with the prior release.

**Comment by Grmiade** [comment]
I know this PR is not mergeable this next days ^^
I'm preparing the future 😄

File: types/jquery/index.d.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@
 //                 Dick van den Brink <https://github.com/DickvdBrink>
 //                 Thomas Schulz <https://github.com/King2500>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
+// TypeScript Version: 2.8
```
Comment (leonard-thieu) [inline types/jquery/index.d.ts]: jQuery does not have a dependency on React. This version bump should not be necessary.

**Review by leonard-thieu — CHANGES_REQUESTED** [review]

**Review by tomwanzek — CHANGES_REQUESTED** [review]
Neither `d3` nor any of the `d3-xxx` module definitions have any dependency on React, so beyond the general point made by @DanielRosenwasser, these definitions do not require a TS 2.8, even if one addressed the React part.

File: types/jquery/index.d.ts (hunk 1 above)
Comment (Grmiade) [reply types/jquery/index.d.ts]: react => react-mce => tinymce => jquery 😞 
Normally, I updated only packages where a dependency error was raised.
If you have a better solution, I take it 😉

**Comment by Grmiade** [comment]
> react => react-mce => tinymce => jquery 😞
> Normally, I updated only packages where a dependency error was raised.
> If you have a better solution, I take it 😉

@tomwanzek For d3, the dependency path is react => react-faux-dom => d3

File: types/jquery/index.d.ts (hunk 1 above)
Comment (leonard-thieu) [reply types/jquery/index.d.ts]: > react => react-mce => tinymce => jquery :disappointed:

This is the actual dependency graph. Neither `@types/jquery` or `@types/tinymce` should require the version bump.

chrome_2018-03-28_14-57-17

Could you double check that jQuery indeed causes the error without the version bump?

It should be the case that only dependents of React need to be updated, not dependencies (@andy-ms ?). 

If it is the case that jQuery requires the version bump, I'd be against this change even when 2.8 becomes allowed on DefinitelyTyped. Increasing requirements soon after a new version becomes allowed seems to cause breaks for a lot of jQuery consumers. It'd be prudent to wait even longer to minimize this.

Comment (tomwanzek) [reply types/jquery/index.d.ts]: Same as the comment above by @leonard-thieu with respect to `d3` et al. None of the D3 modules has an external dependency outside the D3 ecosystem, the only exception is `GeoJson` for a couple of module _definitions_. There is certainly no dependency on `react` or `react-faux-dom`.

As for our current versioning policy, we are migrating the definitions to start using some TS 2.3 features, as the release is ~1yr out and the installed based should be able to adopt readily. TS 2.8 is likely a while out, regardless of how beneficial the new features may be.

Comment (Grmiade) [reply types/jquery/index.d.ts]: Yes I confirm @leonard-thieu 
I understand your remarks.
This dependency system is blocking :/ Conditional types feature is sooo powerful to fix some issues.

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.

**Pull request closed without merging**